### PR TITLE
feat: Machine option + JSON synonyms validation

### PR DESCRIPTION
**Pull request by jirimosinger**

### CODE REVIEW

File: clients/vscode-hlasmplugin/CHANGELOG.md [hunk 1]
```diff
@@ -6,10 +6,11 @@
 - Allow viewing content generated by AINSERT instruction and preprocessors
 - Expand the list of associated file extensions
 - DB2 preprocessor now supports the VERSION option
-- Instruction set versioning support
+- Instruction set versioning support (via OPTABLE and MACHINE options)
 - Basic GOFF, XOBJECT and SYSOPT_XOBJECT support
 - MNOTE support
 - Assembler options can be specified in pgm_conf.json to override values from proc_grps.json
+- User is now prevented from entering synonymous assembler option entries
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/CHANGELOG.md:13]: I don't think we need the second entry, this is a change to an unrelease feature.

File: clients/vscode-hlasmplugin/schema/asm_options.json [hunk 2]
```diff
@@ -1,56 +1,148 @@
...
-                "Z12",
-                "ZS6",
-                "Z13",
-                "ZS7",
-                "Z14",
-                "ZS8",
-                "Z15",
-                "ZS9"
-            ]
+  "type": "object",
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/schema/asm_options.json:2]: reformat?

File: language_server/src/lsp/feature_workspace_folders.cpp [hunk 3]
```diff
@@ -19,6 +19,7 @@
 #include "network/uri/uri.hpp"
 
 #include "../logger.h"
+#include "config/pgm_conf.h"
```
Comment (slavek-kucera) [inline language_server/src/lsp/feature_workspace_folders.cpp:22]: is this change still needed?

File: parser_library/src/instruction_set_version.h [hunk 4]
```diff
@@ -39,10 +39,29 @@ enum class instruction_set_version
     UNI
 };
 
-constexpr std::array<std::pair<std::string_view, instruction_set_version>, 23> instr_set_version_equivalents = { {
+constexpr std::array<std::pair<std::string_view, instruction_set_version>, 77> instr_set_version_equivalents = { {
```
Comment (slavek-kucera) [inline parser_library/src/instruction_set_version.h:42]: is using machine args in optable and vice versa actually supported?

could the definition be moved to a cpp file as a constinit extern const?

File: parser_library/src/config/assembler_options.cpp [hunk 5]
```diff
@@ -116,39 +116,58 @@ void to_json(nlohmann::json& j, const assembler_options& p)
...
-    optional_from_json(j, "SYSTEM_ID", p.system_id);
+    for (const auto& item : j.items())
+    {
+        const auto& key = item.key();
+
+        if (auto it = from_json_map.find(key); it != from_json_map.end())
+        {
+            it->second(item.value(), p);
+        }
+    }
```
Comment (slavek-kucera) [inline parser_library/src/config/assembler_options.cpp:171]: I think that the original solution with a function call per field was both more compact and more readable.
Especially after the change to optional_from_json there is even no reason for the extra goff handling routine.

File: parser_library/test/workspace_manager_test.cpp [hunk 6]
```diff
@@ -16,6 +16,7 @@
 
 #include "gtest/gtest.h"
 
+#include "config/pgm_conf.h"
```
Comment (slavek-kucera) [inline parser_library/test/workspace_manager_test.cpp:19]: not needed?

File: clients/vscode-hlasmplugin/CHANGELOG.md (hunk 1 above)
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/CHANGELOG.md:13]: Ok

File: clients/vscode-hlasmplugin/schema/asm_options.json (hunk 2 above)
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/schema/asm_options.json:2]: Yeah, it is formatted. Maybe we have different formatting settings?

File: language_server/src/lsp/feature_workspace_folders.cpp (hunk 3 above)
Comment (jirimosinger) [reply language_server/src/lsp/feature_workspace_folders.cpp:22]: Yes, because of the call on line 166 where `load_from_json` is called

File: parser_library/src/config/assembler_options.cpp (hunk 5 above)
Comment (jirimosinger) [reply parser_library/src/config/assembler_options.cpp:171]: I also thought about it but the scenario where we are checking the presence of all `assembler_option` members in an almost empty `json` object seemed like a real scenario to me and I didn't want to waste resources on these checks.

If your experience is that user usually provides many assembler options and we will be checking them anyway then I can most certainly change this back.

File: parser_library/test/workspace_manager_test.cpp (hunk 6 above)
Comment (jirimosinger) [reply parser_library/test/workspace_manager_test.cpp:19]: Still needed because of line 113 where `load_from_json` is called

File: language_server/src/lsp/feature_workspace_folders.cpp (hunk 3 above)
Comment (slavek-kucera) [reply language_server/src/lsp/feature_workspace_folders.cpp:22]: Curious, it definitely looks like it needs to be there, I wonder why the build was passing till now...

File: parser_library/src/config/assembler_options.cpp (hunk 5 above)
Comment (slavek-kucera) [reply parser_library/src/config/assembler_options.cpp:171]: I think that performance is not a concern here, but the previous version seems more easily readable to me.

File: language_server/src/lsp/feature_workspace_folders.cpp (hunk 3 above)
Comment (jirimosinger) [reply language_server/src/lsp/feature_workspace_folders.cpp:22]: I think it has something to do with type traits and I've tracked it to this release: https://github.com/nlohmann/json/releases/tag/v3.7.0

Also this issue talks about something very similar too: 
https://github.com/nlohmann/json/issues/1996

Comment (slavek-kucera) [reply language_server/src/lsp/feature_workspace_folders.cpp:22]: ok, I think I see what is happening... there must be some new conversion operator that is trying to convert the json to the pgm_conf structure... isn't there an option to disable implicit json->anything conversion?
The config/pgm_conf.h is the library internal include, so we should not use it in the server source directly.
Renaming the method to load_from_pgm_config seems preferable to prevent the conversion attempt.

Comment (jirimosinger) [reply language_server/src/lsp/feature_workspace_folders.cpp:22]: I didn't find any suitable option -> function renamed

File: parser_library/src/instruction_set_version.h (hunk 4 above)
Comment (jirimosinger) [reply parser_library/src/instruction_set_version.h:42]: 1: Not it's not supported -> added machine variable

2: Moved to `assembler_options.cpp` and left as `constexpr`

File: parser_library/src/config/assembler_options.cpp (hunk 5 above)
Comment (jirimosinger) [reply parser_library/src/config/assembler_options.cpp:171]: Ok

File: parser_library/src/config/assembler_options.cpp [hunk 7]
```diff
@@ -14,23 +14,134 @@
...
+#endif
+
+void to_upper(std::string& s)
+{
+    for (auto& c : s)
+        c = static_cast<char>(std::toupper((unsigned char)c));
+}
+
+template<typename INSTR_SET_EQUIVALENTS>
+bool instr_set_equivalent_valid(std::string instr_set_name, const INSTR_SET_EQUIVALENTS& equivalents) noexcept
```
Comment (slavek-kucera) [inline parser_library/src/config/assembler_options.cpp:133]: could be just std::span...

File: parser_library/test/workspace/load_config_test.cpp [hunk 8]
```diff
@@ -354,4 +354,102 @@ TEST(workspace, asm_options_invalid)
...
+    {}
+
+    const std::string& get_text() override { return proc_file; }
+
+    std::string proc_file;
+
+private:
+    std::string generate_proc_file(std::string_view asm_opts_contents)
+    {
+        return is_windows() ?
```
Comment (slavek-kucera) [inline parser_library/test/workspace/load_config_test.cpp:375]: I don't see any path separators in these inputs, so no need to have win/non-win version.

File: parser_library/src/config/assembler_options.cpp [hunk 9]
```diff
@@ -100,55 +223,43 @@ template<typename T>
...
+    optional_from_json(j, "GOFF", p.goff);
+    optional_from_json(j, "MACHINE", p.machine);
     optional_from_json(j, "OPTABLE", p.optable);
     optional_from_json(j, "PROFILE", p.profile);
     optional_from_json(j, "SYSPARM", p.sysparm);
     optional_from_json(j, "SYSTEM_ID", p.system_id);
+    optional_from_json(j, "XOBJECT", p.goff);
+
+    // Machine and OPTABLE are basically synonyms
+    if (p.machine.has_value() && p.optable.has_value())
```
Comment (slavek-kucera) [inline parser_library/src/config/assembler_options.cpp:260]: assembler_options::valid() should catch this situation, right?

Comment (jirimosinger) [reply parser_library/src/config/assembler_options.cpp:260]: It does. Nevertheless, it depends on what we want to tell the user in VSCode. The current code in PR results in the following when both options are used:

image


Without the check it results in this:
image


Personally, I prefer the first diag message as the option isn't exactly invalid. What do you think?

Comment (slavek-kucera) [reply parser_library/src/config/assembler_options.cpp:260]: I think the second one is more accurate. Additionally, it mentions the name of the processor group, that the problem is related to assembler options.

**Review by slavek-kucera — APPROVED** [review]
